ccl/changefeedccl: TestChangefeedOnlyInitialScan failed

ccl/changefeedccl.TestChangefeedOnlyInitialScan [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8208744?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8208744?buildTab=artifacts#/) on master @ [642afd6e8c8d1f967da4aa0c3c08e2bdc0495100](https://github.com/cockroachdb/cockroach/commits/642afd6e8c8d1f967da4aa0c3c08e2bdc0495100):


```
=== RUN   TestChangefeedOnlyInitialScan
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/905465f697dc04c39f7d8a82cb057bfe/logTestChangefeedOnlyInitialScan2786422312
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestChangefeedOnlyInitialScan
    changefeed_test.go:6568: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/905465f697dc04c39f7d8a82cb057bfe/logTestChangefeedOnlyInitialScan2786422312
--- FAIL: TestChangefeedOnlyInitialScan (85.41s)
=== RUN   TestChangefeedOnlyInitialScan/enterprise
    helpers_test.go:751: making server as secondary tenant
    helpers_test.go:837: making enterprise feed factory
    helpers_test.go:851: pgURL enterprise root
    --- FAIL: TestChangefeedOnlyInitialScan/enterprise (85.39s)
=== RUN   TestChangefeedOnlyInitialScan/enterprise/initial_backfill_only
    changefeed_test.go:6557: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3308/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/changefeedccl/changefeedccl_test_/changefeedccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/changefeedccl/changefeed_test.go:6557
        	Error:      	Not equal: 
        	            	expected: 5000
        	            	actual  : 0
        	Test:       	TestChangefeedOnlyInitialScan/enterprise/initial_backfill_only
        --- FAIL: TestChangefeedOnlyInitialScan/enterprise/initial_backfill_only (42.79s)
=== RUN   TestChangefeedOnlyInitialScan/enterprise/initial_scan_only
    changefeed_test.go:6557: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3308/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/changefeedccl/changefeedccl_test_/changefeedccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/changefeedccl/changefeed_test.go:6557
        	Error:      	Not equal: 
        	            	expected: 5000
        	            	actual  : 0
        	Test:       	TestChangefeedOnlyInitialScan/enterprise/initial_scan_only
        --- FAIL: TestChangefeedOnlyInitialScan/enterprise/initial_scan_only (36.26s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #92887 ccl/changefeedccl: TestChangefeedOnlyInitialScan failed [C-test-failure O-robot T-cdc branch-release-22.2.0]
- #92811 ccl/changefeedccl: TestChangefeedOnlyInitialScan failed [C-test-failure O-robot T-cdc branch-release-22.2]
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedOnlyInitialScan.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-23155

ccl/changefeedccl.TestChangefeedOnlyInitialScan [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8223913?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8223913?buildTab=artifacts#/) on master @ [7594fb2d1438dd5ed8f80a887b8d60b6cdd296ed](https://github.com/cockroachdb/cockroach/commits/7594fb2d1438dd5ed8f80a887b8d60b6cdd296ed):


```
=== RUN   TestChangefeedOnlyInitialScan
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/905465f697dc04c39f7d8a82cb057bfe/logTestChangefeedOnlyInitialScan43750334
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestChangefeedOnlyInitialScan
    changefeed_test.go:6568: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/905465f697dc04c39f7d8a82cb057bfe/logTestChangefeedOnlyInitialScan43750334
--- FAIL: TestChangefeedOnlyInitialScan (87.62s)
=== RUN   TestChangefeedOnlyInitialScan/enterprise
    helpers_test.go:751: making server as secondary tenant
    helpers_test.go:837: making enterprise feed factory
    helpers_test.go:851: pgURL enterprise root
    --- FAIL: TestChangefeedOnlyInitialScan/enterprise (87.61s)
=== RUN   TestChangefeedOnlyInitialScan/enterprise/initial_backfill_only
    changefeed_test.go:6557: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/4148/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/changefeedccl/changefeedccl_test_/changefeedccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/changefeedccl/changefeed_test.go:6557
        	Error:      	Not equal: 
        	            	expected: 5000
        	            	actual  : 0
        	Test:       	TestChangefeedOnlyInitialScan/enterprise/initial_backfill_only
        --- FAIL: TestChangefeedOnlyInitialScan/enterprise/initial_backfill_only (43.78s)
=== RUN   TestChangefeedOnlyInitialScan/enterprise/initial_scan_only
    changefeed_test.go:6557: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/4148/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/ccl/changefeedccl/changefeedccl_test_/changefeedccl_test.runfiles/com_github_cockroachdb_cockroach/pkg/ccl/changefeedccl/changefeed_test.go:6557
        	Error:      	Not equal: 
        	            	expected: 5000
        	            	actual  : 0
        	Test:       	TestChangefeedOnlyInitialScan/enterprise/initial_scan_only
        --- FAIL: TestChangefeedOnlyInitialScan/enterprise/initial_scan_only (37.62s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #92887 ccl/changefeedccl: TestChangefeedOnlyInitialScan failed [C-test-failure O-robot T-cdc branch-release-22.2.0]
- #92811 ccl/changefeedccl: TestChangefeedOnlyInitialScan failed [C-test-failure O-robot T-cdc branch-release-22.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedOnlyInitialScan.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -6517,54 +6517,41 @@ func TestChangefeedOnlyInitialScan(t *testing.T) {
 			t.Run(testName, func(t *testing.T) {
 				sqlDB.Exec(t, `CREATE TABLE foo (a INT PRIMARY KEY)`)
 				sqlDB.Exec(t, `INSERT INTO foo (a) SELECT * FROM generate_series(1, 5000);`)
-
-				// Most changefeed tests can afford to have a race condition between the initial
-				// inserts and starting the feed because the output looks the same for an initial
-				// scan and an insert. For tests with initial_scan=only, though, we can't start the feed
-				// until it's going to see all the initial inserts in the initial scan.
-				sqlDB.CheckQueryResultsRetry(t, `SELECT count(*) FROM foo`, [][]string{{`5000`}})
+				defer func() {
+					sqlDB.Exec(t, `DROP TABLE foo`)
+				}()
 
 				feed := feed(t, f, changefeedStmt)
+				defer closeFeed(t, feed)
 
+				// Insert few more rows after the feed started -- we should not see those emitted.
 				sqlDB.Exec(t, "INSERT INTO foo VALUES (5005), (5007), (5009)")
 
-				g := ctxgroup.WithContext(context.Background())
 				var expectedMessages []string
 				for i := 1; i <= 5000; i++ {
 					expectedMessages = append(expectedMessages, fmt.Sprintf(
 						`foo: [%d]->{"after": {"a": %d}}`, i, i,
 					))
 				}
-				var seenMessages []string
-				g.Go(func() error {
-					for {
-						m, err := feed.Next()
-						if err != nil {
-							return err
-						}
-						seenMessages = append(seenMessages, fmt.Sprintf(`%s: %s->%s`, m.Topic, m.Key, m.Value))
-					}
-				})
 
+				assertPayloads(t, feed, expectedMessages)
+
+				// It would be nice to assert that after we've seen expectedMessages,
+				// that none of the unexpected messages show up before job termination.
+				// However, if any of those unexpected messages were emitted, then, we
+				// would expect this test to flake (hopefully, with an error message
+				// that makes it clear that the unexpected event happen).
 				jobFeed := feed.(cdctest.EnterpriseTestFeed)
 				require.NoError(t, jobFeed.WaitForStatus(func(s jobs.Status) bool {
 					return s == jobs.StatusSucceeded
 				}))
-
-				closeFeed(t, feed)
-				sqlDB.Exec(t, `DROP TABLE foo`)
-				_ = g.Wait()
-				require.Equal(t, len(expectedMessages), len(seenMessages))
-				sort.Strings(expectedMessages)
-				sort.Strings(seenMessages)
-				for i := range expectedMessages {
-					require.Equal(t, expectedMessages[i], seenMessages[i])
-				}
 			})
 		}
 	}
 
-	cdcTest(t, testFn, feedTestEnterpriseSinks)
+	// "enterprise" and "webhook" sink implementations are too slow
+	// for a test that reads 5k messages.
+	cdcTest(t, testFn, feedTestEnterpriseSinks, feedTestOmitSinks("enterprise", "webhook"))
 }
 
 func TestChangefeedOnlyInitialScanCSV(t *testing.T) {

